protocol/proto/addr_enc_dec.go

```diff
@@ -0,0 +1,254 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package proto
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3x/clock"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/retry"
+
+	"github.com/uber-go/tally"
+	"go.uber.org/atomic"
+)
+
+const (
+	defaultRetryForever = true
+)
+
+var (
+	errNoValidConnection = errors.New("no valid connection available")
+)
+
+type connectFn func(addr string) (net.Conn, error)
+
+type addrEncdec struct {
+	sync.RWMutex
+
+	encdec              ConnectionEncodeDecoder
+	addr                string
+	validConn           *atomic.Bool // Use atomic for r/w fast path.
+	resetCh             chan struct{}
+	resetNanos          int64
+	retrier             retry.Retrier
+	dialTimeout         time.Duration
+	reconnectDelayNanos int64
+	doneCh              chan struct{}
+	closed              bool
+	logger              log.Logger
+	m                   *addrEncdecMetrics
+
+	nowFn     clock.NowFn
+	connectFn connectFn
+}
+
+type addrEncdecMetrics struct {
+	resetConn      tally.Counter
+	resetConnError tally.Counter
+}
+
+// NewAddressEncodeDecoder creates an EncodeDecoder that
+// manages the connection to the target address.
+func NewAddressEncodeDecoder(
+	addr string,
+	opts AddressEncodeDecoderOptions,
+) AddressEncodeDecoder {
+	if opts == nil {
+		opts = NewAddressEncodeDecoderOptions()
+	}
+	c := addrEncdec{
+		encdec:     NewConnectionEncodeDecoder(nil, opts.ConnectionEncodeDecoderOptions()),
+		addr:       addr,
+		validConn:  atomic.NewBool(false),
+		resetCh:    make(chan struct{}, 1),
+		resetNanos: time.Now().UnixNano(),
+		// Should retry forever on reconnect.
+		retrier:             retry.NewRetrier(opts.ConnectionRetryOptions().SetForever(defaultRetryForever)),
+		dialTimeout:         opts.DialTimeout(),
+		reconnectDelayNanos: opts.ReconnectDelay().Nanoseconds(),
```

Comment:
Hm should this be part of the retry options?

Reply:
This is a different delay, it avoids signaling connection reset while a reset is ongoing, it's a nice-to-have, not required.